code/datums/achievements/_awards.dm

```diff
@@ -217,3 +233,114 @@
+ * The value of the score should equal to the length of objects that have been achieved.
+ * It also has an unique tab in the UI that lets you review the progress.
+ */
+/datum/award/score/progress
+	///The id of table we're looking for, the one containing the progress objects.
+	var/table_id = ""
+	/**
+	 * When get_changed_rows is called, this list gets filled with the entries to be mass-inserted
+	 * in the table associated with this progress score.
+	 */
+	var/list/changed_entries = list()
+
+/datum/award/score/progress/New()
+	. = ..()
+	RegisterSignal(SSachievements, COMSIG_ACHIEVEMENTS_SAVED_TO_DB, PROC_REF(insert_entries))
+
+/datum/award/score/progress/unlock(mob/user, datum/achievement_data/holder, value)
+	var/list/entries = holder.data[type]
+	if(!value)
+		CRASH("empty value used as argument to progress this score award.")
+	if(value in entries)
+		return
+	// This ensures that the original list and the new won't be the same any longer.
+	// So that it'll pass the not-equal if statement and be saved in the db.
+	if(entries == holder.original_cached_data[type])
+		entries = entries?.Copy() || list()
+		holder.data[type] = entries
+	entries |= value
+
+/datum/award/score/progress/load(datum/achievement_data/holder)
+	var/list/results = ..()
+	return validate_loaded_data(holder, results)
+
+/datum/award/score/progress/on_achievement_data_init(datum/achievement_data/holder, database_value)
+	var/list/results = parse_value(get_raw_value(holder.owner_ckey))
+	validate_loaded_data(holder, results)
+
+/datum/award/score/progress/proc/validate_loaded_data(datum/achievement_data/holder, list/results)
+	holder.original_cached_data[type] = holder.data[type] = results
+	if(!length(results))
+		return results
+	///This list will be populated on validate_entries()
+	var/list/validated_results = list()
+	if(!validate_entries(results, validated_results))
+		holder.data[type] = validated_results
+	return validated_results
+
+///Along with the changed rows for the main table, this also populates changed_entries with the entries list
+/datum/award/score/progress/get_changed_rows(datum/achievement_data/holder)
+	if(!database_id || !holder || !name || !table_id)
+		return
+	var/list/entries = holder.data[type]
+	for(var/entry in (entries - holder.original_cached_data[type]))
+		changed_entries += list(list("ckey" = holder.owner_ckey, "progress_entry" = entry))
+	return list(
+		"ckey" = holder.owner_ckey,
+		"achievement_key" = database_id,
+		"value" = length(entries),
+	)
+
+/datum/award/score/progress/get_ui_data(list/award_data, datum/achievement_data/holder)
+	. = ..()
+	award_data["value"] = length(holder.data[type])
+
+///We don't care much about the default value, which is only used for high scores. Instead, we get the entries string from the db.
+/datum/award/score/progress/get_raw_value(key)
+	var/list/entries = list()
+	var/datum/db_query/get_entries_load = SSdbcore.NewQuery(
```

Comment:
this needs an IsAdminAdvancedProcCall in front of it, as do any db calls

Reply:
Db queries already block admin proc calls.

Reply:
https://github.com/tgstation/tgstation/blob/3f50f5eac666652aeb4aa48c3b80549c91380ebd/code/controllers/subsystem/dbcore.dm#L381-L384

`dbcore` is chuck-full of `IsAdminAdvancedProcCall()` checks.

Reply:
thought MassInsert didn't but it seems it does, scratch this then